File: mobile/app/src/main/java/com/praeter/ui/mainactivity/BottomSheetFragment.java
package com.praeter.ui.mainactivity;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.os.Parcelable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.AppCompatImageView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;
import com.praeter.R;
import com.praeter.core.utils.PraeterNetworkManager;
import com.praeter.data.local.model.Class;

import org.parceler.Parcels;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import jp.wasabeef.glide.transformations.BlurTransformation;
import timber.log.Timber;


@SuppressLint("NonConstantResourceId")
public class BottomSheetFragment extends BottomSheetDialogFragment {

    public static final String CLASS_ITEM_BUNDLE = "CLASS_ITEM_BUNDLE";

    private Context context;

    @BindView(R.id.close_bottom_sheet_btn)
    ImageButton btnCloseBottomSheet;
    @BindView(R.id.iv_class_thumb_blurred)
    AppCompatImageView ivThumbBlurred;
    @BindView(R.id.iv_class_thumb)
    AppCompatImageView ivThumb;
    @BindView(R.id.tv_name)
    TextView tvName;
    @BindView(R.id.tv_type)
    TextView tvType;
    @BindView(R.id.btn_book)
    Button btnBook;

    private Class storedClass;


    public static BottomSheetFragment newInstance() {

        Bundle args = new Bundle();

        BottomSheetFragment fragment = new BottomSheetFragment();
        fragment.setArguments(args);
        return fragment;
    }


    public static BottomSheetFragment newInstance(Class classItem) {

        Bundle args = new Bundle();

        Parcelable parcelable = Parcels.wrap(Class.class, classItem);

        args.putParcelable(CLASS_ITEM_BUNDLE, parcelable);

        BottomSheetFragment fragment = new BottomSheetFragment();
        fragment.setArguments(args);
        return fragment;
    }

    public BottomSheetFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View rootView =
                inflater.inflate(
                        R.layout.fragment_bottom_sheet_dialog,
                        container,
                        false);
        ButterKnife.bind(this, rootView);
        return rootView;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Timber.i("onViewCreated()");

        context = getActivity();

        //Retrieve data
        getClassData();
    }


    @OnClick(R.id.btn_book)
    void onReserveButtonClicked() {
        if (!PraeterNetworkManager.isConnected(context)) {
            String errorMessage = "Not connected to the internet. Check your internet connection.";
            Timber.e(errorMessage);
            Toast.makeText(context, errorMessage, Toast.LENGTH_LONG).show();
            return;
        }

        String message = "You've successfully reserved the class " + storedClass.getName();
        Toast.makeText(context, message, Toast.LENGTH_LONG).show();

    }

    @OnClick(R.id.close_bottom_sheet_btn)
    void onCloseButtonClicked() {
        dismiss();
    }

    private void getClassData() {
        Timber.i("getDeviceInfo()");

        assert getArguments() != null;
        if (null == getArguments().getParcelable(CLASS_ITEM_BUNDLE)) {
            Timber.e("Bundle with key : CLASS_ITEM_BUNDLE, is null ");
            return;
        }

        storedClass = Parcels.unwrap(getArguments().getParcelable(CLASS_ITEM_BUNDLE));

        setViews(storedClass);
    }

    @SuppressLint("SetTextI18n")
    private void setViews(Class classItem) {
        Glide.with(this)
                .load(R.drawable.visa_icon)
                .transform(
                        new BlurTransformation(75),
                        new RoundedCorners(48))
                .into(ivThumbBlurred);

        tvName.setText(classItem.getName());
        tvType.setText(classItem.getType());
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        Timber.e("DestroyView()");
    }
}

File: mobile/app/src/main/java/com/praeter/ui/mainactivity/fragment/bottomfragment/BottomSheetFragment.kt
package com.praeter.ui.mainactivity.fragment.bottomfragment

import android.annotation.SuppressLint
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.bitmap.RoundedCorners
import com.google.android.material.bottomsheet.BottomSheetDialogFragment
import com.praeter.R
import com.praeter.core.utils.PraeterNetworkManager
import com.praeter.data.local.model.Ancient
import com.praeter.data.local.model.ClassModel
import com.praeter.databinding.FragmentBottomSheetDialogBinding
import jp.wasabeef.glide.transformations.BlurTransformation
import timber.log.Timber

class BottomSheetFragment : BottomSheetDialogFragment() {

    private var _viewBinding: FragmentBottomSheetDialogBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _viewBinding!!

    private var storedClass: Any? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _viewBinding = FragmentBottomSheetDialogBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        Timber.i("onViewCreated()")

        //Retrieve data
        classData
    }

    override fun onDestroyView() {
        super.onDestroyView()

        _viewBinding = null
    }

    fun book(view: View) {
        Timber.i("onReserveButtonClicked()")
        if (!PraeterNetworkManager.isConnected(requireContext())) {
            val errorMessage = "Not connected to the internet. Check your internet connection."
            Timber.e(errorMessage)
            Toast.makeText(context, errorMessage, Toast.LENGTH_LONG).show()
            return
        }
        val message = "You've successfully reserved the class " +
                if (storedClass is ClassModel) (storedClass as ClassModel?)?.name else (storedClass as Ancient?)?.name
        Toast.makeText(context, message, Toast.LENGTH_LONG).show()
    }

    fun close(view: View) {
        Timber.i("onCloseButtonClicked()")
        dismiss()
    }

    private val classData: Unit
        get() {
            Timber.i("getClassData()")
            assert(arguments != null)

            if (null == arguments?.getParcelable(CLASS_ITEM_BUNDLE)) {
                Timber.e("Bundle with key : CLASS_ITEM_BUNDLE, is null ")
                return
            }

            storedClass = arguments?.getParcelable(CLASS_ITEM_BUNDLE)

            if (storedClass is ClassModel) setViews((storedClass as ClassModel?)!!) else if (storedClass is Ancient) setViews(
                (storedClass as Ancient?)!!
            )
        }

    @SuppressLint("SetTextI18n")
    private fun setViews(classModelItem: ClassModel) {
        Glide.with(this)
            .load(R.drawable.visa_icon)
            .transform(
                BlurTransformation(75),
                RoundedCorners(48)
            )
            .into(binding.ivClassThumbBlurred)

        binding.tvName.text = classModelItem.name
        binding.tvType.text = classModelItem.type
    }

    @SuppressLint("SetTextI18n")
    private fun setViews(ancientItem: Ancient) {
        Glide.with(this)
            .load(R.drawable.visa_icon)
            .transform(
                BlurTransformation(75),
                RoundedCorners(48)
            )
            .into(binding.ivClassThumbBlurred)

        binding.tvName.text = ancientItem.name
        binding.tvType.text = ancientItem.name
    }

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)
        Timber.e("onDismiss()")
    }

    companion object {
        const val CLASS_ITEM_BUNDLE = "CLASS_ITEM_BUNDLE"
        fun newInstance(): BottomSheetFragment {
            val args = Bundle()
            val fragment = BottomSheetFragment()
            fragment.arguments = args
            return fragment
        }

        fun newInstance(classModelItem: ClassModel?): BottomSheetFragment {
            val args = Bundle()
            args.putParcelable(CLASS_ITEM_BUNDLE, classModelItem)
            val fragment = BottomSheetFragment()
            fragment.arguments = args
            return fragment
        }

        fun newInstance(ancientItem: Ancient?): BottomSheetFragment {
            val args = Bundle()
            args.putParcelable(CLASS_ITEM_BUNDLE, ancientItem)
            val fragment = BottomSheetFragment()
            fragment.arguments = args
            return fragment
        }
    }
}